django/db/backends/oracle/operations.py

```diff
@@ -699,6 +706,12 @@ def subtract_temporals(self, internal_type, lhs, rhs):
 
     def bulk_batch_size(self, fields, objs):
         """Oracle restricts the number of parameters in a query."""
+        fields = list(
+            chain.from_iterable(
+                field.fields if isinstance(field, CompositePrimaryKey) else [field]
+                for field in fields
+            )
+        )
```

Comment:
We definitely want to add this for now due to `TupleExact.as_oracle` but one thing that I noticed while playing with the docker-compose version is that Oracle 23ai (actually 23c renamed to 23ai because why not 😅 ) appears to support native tuple comparison while 19c still doesn't for most operations?

One thing that is very weird though is that [I can't confirm from raw SQL](https://dbfiddle.uk/SCfB7ZKF) so I wonder if the `oracledb` wrapper does something special?

Anyway, I'll continue investigating and this should go in as-is for now.

Reply:
See #19107 as the conclusion of this investigation.